contrib/storage-druid/src/main/java/org/apache/drill/exec/store/druid/DruidRecordReader.java

```diff
@@ -0,0 +1,193 @@
+  private final DruidSubScan.DruidSubScanSpec scanSpec;
+  private final List<String> dimensions;
+  private final DruidFilter filter;
+  private ArrayList<PagingIdentifier> pagingIdentifiers = new ArrayList<>();
+  private int maxRecordsToRead = -1;
+
+  private JsonReader jsonReader;
+  private VectorContainerWriter writer;
+
+  private final FragmentContext fragmentContext;
+
+  public DruidRecordReader(DruidSubScan.DruidSubScanSpec subScanSpec,
+                           List<SchemaPath> projectedColumns,
+                           int maxRecordsToRead,
+                           FragmentContext context,
+                           DruidStoragePlugin plugin) {
+    dimensions = new ArrayList<>();
+    setColumns(projectedColumns);
+    this.maxRecordsToRead = maxRecordsToRead;
+    this.plugin = plugin;
+    scanSpec = subScanSpec;
+    fragmentContext = context;
+    this.filter = subScanSpec.getFilter();
+  }
+
+  @Override
+  protected Collection<SchemaPath> transformColumns(Collection<SchemaPath> projectedColumns) {
+    Set<SchemaPath> transformed = Sets.newLinkedHashSet();
+    if (isStarQuery()) {
+      transformed.add(SchemaPath.STAR_COLUMN);
+    } else {
+      for (SchemaPath column : projectedColumns) {
+        String fieldName = column.getRootSegment().getPath();
+        transformed.add(column);
+        this.dimensions.add(fieldName);
+      }
+    }
+    return transformed;
+  }
+
+  @Override
+  public void setup(OperatorContext context, OutputMutator output) {
+    this.writer = new VectorContainerWriter(output);
+
+    this.jsonReader =
+      new JsonReader.Builder(fragmentContext.getManagedBuffer())
+        .schemaPathColumns(ImmutableList.copyOf(getColumns()))
+        .skipOuterList(true)
+        .build();
+  }
+
+  @Override
+  public int next() {
+    writer.allocate();
+    writer.reset();
+    DruidQueryClient druidQueryClient = plugin.getDruidQueryClient();
+    Stopwatch watch = Stopwatch.createStarted();
+    try {
+      DruidSelectResponse druidSelectResponse = druidQueryClient.executeQuery(getQuery());
+      setNextPagingIdentifiers(druidSelectResponse);
+
+      int docCount = 0;
+      for (ObjectNode eventNode : druidSelectResponse.getEvents()) {
+        writer.setPosition(docCount);
+        jsonReader.setSource(eventNode);
+        try {
+          jsonReader.write(writer);
+        } catch (IOException e) {
+          String msg = "Failure while reading document. - Parser was at record: " + eventNode.toString();
+          logger.error(msg, e);
+          throw new DrillRuntimeException(msg, e);
+        }
+        docCount++;
+      }
+
+      writer.setValueCount(docCount);
+      logger.debug("Took {} ms to get {} records", watch.elapsed(TimeUnit.MILLISECONDS), docCount);
+      return docCount;
+    } catch (Exception e) {
+      String msg = "Failure while reading documents";
+      logger.error(msg, e);
+      throw new DrillRuntimeException(msg, e);
```

Comment:
Here and above, 
Can you please throw a `UserException` rather than a `DrillRuntimeException`?

Something like this:
```
 throw UserException
        .dataReadError(e)
        .message(msg)
        .build(logger);
```
If you can access the `errorContext` it is also helpful (but not necessary) to add that as well via the `addContext()` method.  That way the user gets helpful error messages if something goes wrong.

Reply:
Fixed.